packages/celotool/src/cmds/bots/auto-vote.ts

```diff
@@ -0,0 +1,318 @@
+import { getAccountAddressFromPrivateKey } from '@celo/walletkit'
+import BigNumber from 'bignumber.js'
+import { groupBy, mapValues } from 'lodash'
+import { envVar, fetchEnv } from 'src/lib/env-utils'
+import { AccountType, getPrivateKeysFor } from 'src/lib/generate_utils'
+import { ensure0x } from 'src/lib/utils'
+import { Argv } from 'yargs'
+
+export const command = 'auto-vote'
+
+export const describe = 'for each of the voting bot accounts, vote for the best groups available'
+
+interface SimulateVotingArgv {
+  celoProvider: string
+}
+
+export const builder = (yargs: Argv) => {
+  return yargs.option('celoProvider', {
+    type: 'string',
+    description: 'The node to use',
+    default: 'http://localhost:8545',
+  })
+}
+
+export const handler = async function simulateVoting(argv: SimulateVotingArgv) {
+  try {
+    const mnemonic = fetchEnv(envVar.MNEMONIC)
+    const numBotAccounts = parseInt(fetchEnv(envVar.VOTING_BOTS), 10)
+
+    const kit = newKit(argv.celoProvider)
+    const election = await kit.contracts.getElection()
+
+    const wakeProbability = new BigNumber(fetchEnv(envVar.VOTING_BOT_WAKE_PROBABILITY))
+    const baseChangeProbability = new BigNumber(fetchEnv(envVar.VOTING_BOT_CHANGE_BASELINE))
+    const exploreProbability = new BigNumber(fetchEnv(envVar.VOTING_BOT_EXPLORE_PROBABILITY))
+    const scoreSensitivity = new BigNumber(fetchEnv(envVar.VOTING_BOT_SCORE_SENSITIVITY))
+
+    const allBotKeys = getPrivateKeysFor(AccountType.VOTING_BOT, mnemonic, numBotAccounts)
+    await activatePendingVotes(kit, allBotKeys)
+
+    const botKeysVotingThisRound = allBotKeys.filter((_) =>
+      wakeProbability.isGreaterThan(Math.random())
+    )
+    console.info(`Participating this time: ${botKeysVotingThisRound.length} of ${numBotAccounts}`)
+
+    // If no bots are participating, return early
+    if (botKeysVotingThisRound.length === 0) {
+      return
+    }
+
+    const groupCapacities = await calculateInitialGroupCapacities(kit)
+    const groupScores = await calculateGroupScores(kit)
+    const groupWeights = calculateGroupWeights(groupScores, scoreSensitivity)
+
+    const unelectedGroups = Object.keys(groupCapacities).filter((k) => !groupScores.has(k))
+
+    for (const key of botKeysVotingThisRound) {
+      const botAccount = ensure0x(getAccountAddressFromPrivateKey(key))
+
+      kit.addAccount(key)
+      kit.defaultAccount = botAccount
+
+      console.info(`Voting as: ${botAccount}.`)
+      try {
+        // Get current vote for this bot.
+        // Note: though this returns an array, the bot process only ever chooses one group,
+        // so this takes a shortcut and only looks at the first in the response
+        const currentVote = (await election.getVoter(botAccount)).votes[0]
+        const currentGroup = currentVote ? currentVote.group : undefined
+
+        // Handle the case where the group the bot is currently voting for does not have a score
+        if (
+          !currentGroup ||
```

Comment:
I'm concerned the bots first votes could decimate the incumbency.  (At least if they're run with enough gold, say 25k.  Otherwise I'm concerned there will be little impact.)

To avoid maybe:
- Run bots once with VOTING_BOT_EXPLORE_PROBABILITY=0, long enough that all of the bots are voting for the incumbency.
- Restart bots with VOTING_BOT_EXPLORE_PROBABILITY=1, so that any displacement of incumbency is based on score.

Reply:
That could work. I'm not sure setting `VOTING_BOT_EXPLORE_PROBABILITY=1` is the best idea. I think having _some_ bots voting based on good performance is useful. Maybe like.... 0.8?